pyanaconda/modules/subscription/runtime.py

```diff
@@ -114,3 +167,121 @@ def run(self):
 
         # and finally set the dict to RHSM via the DBus API
         self._rhsm_config_proxy.SetAll(config_dict, "")
+
+
+class RegisterWithUsernamePasswordTask(Task):
+    """Register the system via username + password."""
+
+    def __init__(self, rhsm_register_server_proxy, username, password):
+        """Create a new registration task.
+
+        It is assumed the username and password have been
+        validated before this task has been started.
+
+        :param rhsm_register_server_proxy: DBus proxy for the RHSM RegisterServer object
+        :param str username: Red Hat account username
+        :param str password: Red Hat account password
+        """
+        super().__init__()
+        self._rhsm_register_server_proxy = rhsm_register_server_proxy
+        self._username = username
+        self._password = password
+
+    @property
+    def name(self):
+        return "Register with Red Hat account username and password"
+
+    def run(self):
+        """Register the system with Red Hat account username and password.
+
+        :return: a (success/failure, error_message) tuple
+        :rtype: a (bool, str) tuple
+        """
+        log.debug("subscription: connecting to RHSM private bus")
+        private_register_proxy = start_rhsm_private_bus(self._rhsm_register_server_proxy)
+        log.debug("subscription: registering with username and password")
+        success = False
+        error_message = ""
+        try:
+            locale = os.environ.get("LANG", "")
+            # We do not yet support setting organization for username & password
+            # registration, so organization is blank for now.
+            organization = ""
+            result = private_register_proxy.Register(organization,
+                                                     self._username,
+                                                     self._password,
+                                                     {},
+                                                     {},
+                                                     locale)
+            self.subscription_json = result
+            success = True
+            log.debug("subscription: registered with username and password")
+        except DBusError as e:
+            log.debug("subscription: failed to register with username and password: %s", str(e))
+            # RHSM exception contain details as JSON due to DBus exception handling limitations
+            exception_dict = json.loads(str(e))
+            # return a generic error message in case the RHSM provided error message is missing
+            message = exception_dict.get("message", _("Registration failed."))
+            error_message = message
+        finally:
+            # registration done (one way or another), so shutdown the private bus used for
+            # credential passing
+            stop_rhsm_private_bus(self._rhsm_register_server_proxy)
+        # return the result
+        return success, error_message
```

Comment:
Why don't you raise an exception?

Reply:
I was not sure we can safely survive an exception being raised in the task, so opted for the tuple for now. The thing is that the registration task will fail quite often in practice due to things such as typos in usernames and passwords & Anaconda needs to be able to survive that, notify user & try again.

If it is not a problem to safely catch the exception (mostly likely from the `run_sync()` method that runs the task in the registration handling thread in the UI) the we should definitely use an exception for this. :)

Reply:
It is definitely safe, we use it a lot :-) The task's exception is propagated via DBus when we call it's `Finish` method in the function `sync_run_task`. So you can safely catch it on the client side when you call  `sync_run_task`.

Reply:
Sounds good, I'll switch the tasks to return exceptions then. :)